File: app/src/main/java/com/xianzhitech/ptt/service/SocketIOProvider.java
package com.xianzhitech.ptt.service;

import android.support.annotation.NonNull;
import android.util.Base64;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Iterables;
import com.xianzhitech.model.Group;
import com.xianzhitech.model.Person;
import com.xianzhitech.ptt.Broker;
import com.xianzhitech.ptt.service.auth.IAuthService;
import com.xianzhitech.ptt.service.signal.Room;
import com.xianzhitech.ptt.service.signal.SignalProvider;
import com.xianzhitech.ptt.service.sync.ISyncService;
import com.xianzhitech.ptt.util.JsonUtil;
import io.socket.client.IO;
import io.socket.client.Manager;
import io.socket.client.Socket;
import io.socket.engineio.client.Transport;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import rx.Observable;
import rx.subjects.PublishSubject;

import java.net.URISyntaxException;
import java.security.MessageDigest;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeoutException;

/**
 *
 * 基于WebSocket的信号服务器
 *
 * Created by fanchao on 13/12/15.
 */
public class SocketIOProvider implements SignalProvider, IAuthService, ISyncService {
    private final String endpointUrl;
    private Socket socket;
    final PublishSubject<Event> eventSubject = PublishSubject.create();
    final Broker broker;

    public static final String EVENT_USER_LOGON = "userLogon";
    public static final String EVENT_CONTACTS_UPDATE = "contactsUpdate";

    public SocketIOProvider(final Broker broker, final String endpointUrl) {
        this.broker = broker;
        this.endpointUrl = endpointUrl;

        eventSubject
                .filter(event -> EVENT_CONTACTS_UPDATE.equals(event.name))
                .flatMap(event -> {
                    final JSONObject object = (JSONObject) event.args[0];
                    final Person person = new Person();
                    final Group group = new Group();
                    try {
                        //TODO: Version control
                        final JSONArray groupJsonArray = object.getJSONObject("enterpriseGroups").getJSONArray("add");
                        final Iterable<Group> groups = JsonUtil.<JSONObject, Group>fromArray(groupJsonArray, group::from);
                        final Iterable<Person> persons = JsonUtil.<JSONObject, Person>fromArray(object.getJSONObject("enterpriseMembers").getJSONArray("add"), person::readFrom);

                        return Observable.concat(
                                broker.updatePersons(persons),
                                broker.updateGroups(groups, GroupMembers.fromJson(groupJsonArray)),
                                broker.updateContacts(Iterables.transform(persons, Person::getId), Iterables.transform(groups, Group::getId))
                        );
                    } catch (JSONException e) {
                        throw new RuntimeException(e);
                    }
                })

                .subscribe();
    }

    @Override
    public Room joinRoom(final String groupId) {
        return null;
    }

    @Override
    public void quitRoom(final String groupId) {
    }

    @Override
    public boolean requestFocus(final int roomId) {
        return false;
    }

    @Override
    public void releaseFocus(final int roomId) {

    }

    @Override
    public void stopSync() {

    }

    @Override
    public Observable<Person> login(final String username, final String password) {
        if (socket != null) {
            throw new IllegalStateException("Already authorized");
        }

        try {
            socket = IO.socket(endpointUrl);
        } catch (URISyntaxException e) {
            throw new RuntimeException(e);
        }

        socket.io().on(Manager.EVENT_TRANSPORT, args -> {
            final Transport transport = (Transport) args[0];
            transport.on(Transport.EVENT_REQUEST_HEADERS, transportArgs -> ((Map<String, List<String>>) transportArgs[0]).put("Authorization",
                    Collections.singletonList("Basic " + encodeCredentials(username, password))));
        });

        socket.on(Socket.EVENT_CONNECT_ERROR, new EventListener(Socket.EVENT_CONNECT_ERROR))
                .on(Socket.EVENT_CONNECT, new EventListener(Socket.EVENT_CONNECT))
                .on(Socket.EVENT_ERROR, new EventListener(Socket.EVENT_ERROR))
                .on(Socket.EVENT_CONNECT_TIMEOUT, new EventListener(Socket.EVENT_CONNECT_TIMEOUT))
                .on(EVENT_USER_LOGON, new EventListener(EVENT_USER_LOGON))
                .on(EVENT_CONTACTS_UPDATE, new EventListener(EVENT_CONTACTS_UPDATE));

        return eventSubject
                .flatMap(event -> {
                    if (Socket.EVENT_CONNECT_ERROR.equals(event.name) || Socket.EVENT_ERROR.equals(event.name)) {
                        return Observable.error(new RuntimeException("Connection error: " + event.args[0]));
                    } else if (Socket.EVENT_CONNECT_TIMEOUT.equals(event.name)) {
                        return Observable.error(new TimeoutException());
                    } else if (EVENT_USER_LOGON.equals(event.name)) {
                        return Observable.just(new Person().readFrom((JSONObject) event.args[0]));
                    } else if (Socket.EVENT_CONNECT.equals(event.name)) {
                        socket.emit("syncContacts", new JSONObject(ImmutableMap.of("enterMemberVersion", 1, "enterGroupVersion", 1)));
                    }

                    return Observable.empty();
                })
                .doOnSubscribe(socket::connect);
    }

    @NonNull
    private String encodeCredentials(final String username, final String password) {
        return Base64.encodeToString((username + ":" + MD5(password)).getBytes(), Base64.NO_WRAP);
    }

    private static String MD5(String md5) {
        try {
            MessageDigest md = java.security.MessageDigest.getInstance("MD5");
            byte[] array = md.digest(md5.getBytes());
            final StringBuilder sb = new StringBuilder();
            for (final byte b : array) {
                sb.append(Integer.toHexString((b & 0xFF) | 0x100).substring(1, 3));
            }
            return sb.toString();
        } catch (java.security.NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void logout() {
        if (socket != null) {
            socket.close();
            socket = null;
        }
    }

    private class EventListener implements Socket.Listener {
        private final String eventName;

        private EventListener(final String eventName) {
            this.eventName = eventName;
        }

        @Override
        public void call(final Object... args) {
            eventSubject.onNext(new Event(eventName, args));
        }
    }

    private static class Event {
        public final String name;
        public final Object[] args;

        private Event(final String name, final Object[] args) {
            this.name = name;
            this.args = args;
        }
    }
}

File: app/src/main/kotlin/com/xianzhitech/service/provider/SocketIOProvider.kt
package com.xianzhitech.service.provider

import com.google.common.collect.ImmutableMap
import com.google.common.collect.Iterables
import com.xianzhitech.ext.*
import com.xianzhitech.model.Group
import com.xianzhitech.model.Person
import com.xianzhitech.model.toGroupsAndMembers
import com.xianzhitech.ptt.Broker
import com.xianzhitech.ptt.service.signal.Room
import io.socket.client.Ack
import io.socket.client.IO
import io.socket.client.Manager
import io.socket.client.Socket
import io.socket.engineio.client.Transport
import org.json.JSONArray
import org.json.JSONObject
import rx.Observable
import rx.subjects.PublishSubject
import java.util.concurrent.TimeoutException

/**
 * Created by fanchao on 17/12/15.
 */
class SocketIOProvider(private val broker: Broker, private val endpoint: String) : SignalProvider, AuthProvider {
    companion object {
        public const val EVENT_SERVER_USER_LOGON = "s_logon"

        public const val EVENT_CLIENT_SYNC_CONTACTS = "c_sync_contact"
        public const val EVENT_CLIENT_CREATE_ROOM = "c_create_room"
    }

    private var socket: Socket? = null
    private val eventSubject = PublishSubject.create<Event>()

    override fun joinRoom(groupId: String): Room {
        throw UnsupportedOperationException()
    }

    override fun quitRoom(groupId: String) {
        throw UnsupportedOperationException()
    }

    override fun requestFocus(roomId: Int): Boolean {
        throw UnsupportedOperationException()
    }

    override fun releaseFocus(roomId: Int) {
        throw UnsupportedOperationException()
    }

    override fun login(username: String, password: String): Person {
        if (socket != null) {
            throw IllegalStateException("Already logon")
        }

        socket = IO.socket(endpoint).let {
            it.io().on(Manager.EVENT_TRANSPORT, { args: Array<Any> ->
                val arg = args[0] as Transport
                arg.on(Transport.EVENT_REQUEST_HEADERS, {
                    (it[0] as MutableMap<String, List<String>>).put("Authorization", listOf("Basic ${(username + ':' + password.toMD5()).toBase64()}"))
                })
            })

            it.subscribeTo(
                    Socket.EVENT_CONNECT,
                    Socket.EVENT_CONNECT_ERROR,
                    Socket.EVENT_CONNECT_TIMEOUT,
                    Socket.EVENT_ERROR,
                    EVENT_SERVER_USER_LOGON)
            it
        }

        return eventSubject
                .flatMap({ event: Event ->
                    when (event.name) {
                        Socket.EVENT_ERROR, Socket.EVENT_CONNECT_ERROR -> Observable.error<Person>(RuntimeException("Connection error: " + event.args))
                        Socket.EVENT_CONNECT_TIMEOUT -> Observable.error<Person>(TimeoutException())
                        Socket.EVENT_CONNECT -> {
                            syncContacts()
                            Observable.empty<Person>()
                        }
                        EVENT_SERVER_USER_LOGON -> Observable.just<Person>(Person().readFrom(event.jsonObject))
                        else -> Observable.empty<Person>()
                    }
                })
                .doOnSubscribe { socket?.connect() }
                .toBlockingFirst()
    }

    override fun logout() {
        throw UnsupportedOperationException()
    }

    fun syncContacts() {
        socket?.emit(EVENT_CLIENT_SYNC_CONTACTS,
                arrayOf(ImmutableMap.of("enterMemberVersion", 1, "enterGroupVersion", 1).toJSONObject()),
                Ack {
                    val result : JSONObject = it[0] as JSONObject
                    val personBuf = Person()
                    var groupBuf = Group()
                    val persons = result.getJSONObject("enterpriseMembers").getJSONArray("add").transform { personBuf.readFrom(it as JSONObject) }
                    val groupJsonArray = result.getJSONObject("enterpriseGroups").getJSONArray("add")
                    val groups = groupJsonArray.transform { groupBuf.readFrom(it as JSONObject) }
                    Observable.concat(
                            broker.updatePersons(persons),
                            broker.updateGroups(groups, groupJsonArray.toGroupsAndMembers()),
                            broker.updateContacts(Iterables.transform(persons, { it.id }), Iterables.transform(groups, { it.id })))
                    .subscribe()
                })
    }

    fun Socket.subscribeTo(vararg events: String): Socket {
        for (event in events) {
            on(event, { this@SocketIOProvider.eventSubject.onNext(Event(event, it)) })
        }
        return this
    }

}

internal data class Event(val name: String, val args: Array<Any>) {
    val jsonObject: JSONObject
        get() = args[0] as JSONObject

    val jsonArray: JSONArray
        get() = args[0] as JSONArray
}
